lib/protobuf/encoder.ex

```diff
@@ -114,16 +114,16 @@ defmodule Protobuf.Encoder do
     repeated_or_not(val, repeated, fn v ->
       v = if is_map, do: struct(prop.type, %{key: elem(v, 0), value: elem(v, 1)}), else: v
       # so that oneof {:atom, v} can be encoded
-      encoded = encode(type, v, [])
-      byte_size = byte_size(encoded)
-      [fnum, encode_varint(byte_size), encoded]
+      encoded = encode(type, v, iolist: true)
+      byte_size = IO.iodata_length(encoded)
+      [fnum | encode_varint(byte_size)] ++ encoded
     end)
   end
 
   defp encode_field(:packed, val, %{type: type, encoded_fnum: fnum}) do
     encoded = Enum.map(val, fn v -> encode_type(type, v) end)
     byte_size = IO.iodata_length(encoded)
-    [fnum, encode_varint(byte_size), encoded]
+    [fnum | encode_varint(byte_size)] ++ encoded
```

Comment:
Won't this potentially be slower than what we had before, considering that what we had before is constant time?

Reply:
Good question. I think the tradeoff here is keeping the iodata more flat and supposedly faster to convert to binary later on versus leaving it more nested but avoiding the supposedly slower `++` operation, even though the first list is quite small. Absolutely no idea which one will perform best 😅. I'll try to microbenchmark these variations too and see if I get any significant results.

Reply:
Got it. I think that I'd go with leaving it as it is now (since there's no clear winner yet) at least in this PR so that there are less changes to affect benchmarks :)

Reply:
Yeah. In fact, this is a "merge" of about 4 or 5 different branches I had, each with a partial improvement (only `float` and `double` as iodata, only `bytes`, only `embedded`, and so on). But it is easy to get carried away and combine some of these without proper benchmarking. Anyway, we can always go back and undo these smaller changes if they prove bad. Do you think it's a good idea to put these into separate commits? I can do that.